### PR TITLE
Update `pyright` to `1.1.366`

### PR DESCRIPTION
**Pull request by Viicos**
## Change Summary

Needs the changes in https://github.com/pydantic/pydantic-core/pull/1323.

Most ignore comments are due to illegal usage of type forms, so I hope we'll have the typeform pep accepted soon so that we can update these parts of the code :)

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/fields.py [hunk 1]
```diff
@@ -373,12 +373,12 @@ class MyModel(pydantic.BaseModel):
             init_var = False
             if annotation is dataclasses.InitVar:
                 init_var = True
-                annotation = Any
+                annotation = typing.cast(Any, Any)
```
Comment (Viicos) [inline pydantic/fields.py:376]: This is a really weird behavior from pyright, where the following would happen:

```python
from typing import Any

var = Any

reveal_type(var)  # Revealed type is "Any"

def func(arg: type): pass

func(var)  # Type error
```

The reason is that `var` is inferred as being the actual `typing.Any` object, and not the `Any` "typing concept" (but pyright displays it the same way).

This also relates to the Typing form PEP, that would be of great help here.

File: pydantic/fields.py [hunk 2]
```diff
@@ -321,7 +321,7 @@ class MyModel(pydantic.BaseModel):
                 new_field_info.metadata = metadata
                 return new_field_info
 
-        return FieldInfo(annotation=annotation, frozen=final or None)
+        return FieldInfo(annotation=annotation, frozen=final or None)  # pyright: ignore[reportArgumentType]
```
Comment (sydney-runkle) [inline pydantic/fields.py:324]: What's the issue with the argument type here?

File: pydantic/types.py [hunk 3]
```diff
@@ -474,7 +474,7 @@ class ConstrainedExample(BaseModel):
         '''
     ```
     """  # noqa: D212
-    return Annotated[
+    return Annotated[  # pyright: ignore[reportReturnType]
```
Comment (sydney-runkle) [inline pydantic/types.py:477]: Same question - why is this needed here?

Comment (sydney-runkle) [reply pydantic/types.py:477]: Maybe @adriangb would know - you're up to date on typing with `Annotated` 🌟

File: pydantic/fields.py (hunk 1 above)
Comment (sydney-runkle) [reply pydantic/fields.py:376]: Hmph, I was definitely confused by this before reading the explanation. Do you think this is something that might change on the pyright end of things any time soon?

**Review by sydney-runkle — COMMENTED** [review]
Hi @Viicos,

Thanks for your work on this. I left a few lingering questions!

File: pydantic/types.py (hunk 3 above)
Comment (sydney-runkle) [reply pydantic/types.py:477]: This seems fine actually, these functions are already so hacky anyways, I'm not surprised this is a problem.

File: pydantic/fields.py (hunk 1 above)
Comment (Viicos) [reply pydantic/fields.py:376]: I'll open a discussion on the pyright repo to see if we can find a better workaround

File: pydantic/fields.py (hunk 2 above)
Comment (Viicos) [reply pydantic/fields.py:324]: `annotation` is inferred as `type[Any] | type[Final]` because of the `if annotation is not typing_extensions.Final:` check on line 300. And with the recent changes in pyright regarding special forms (such as `Final`), `type[Final]` is not compatible with `type[Any]`. Hopefully this will also be resolved with the TypeForm PEP.

File: pydantic/fields.py (hunk 1 above)
Comment (Viicos) [reply pydantic/fields.py:376]: Unfortunately we are using some undefined behavior here (`Any` is meant to be used as an annotation, not as a value), so I don't expect pyright to special case things here :/ But this clearly shows the need for a TypeForm feature in the typing system!

Comment (Viicos) [reply pydantic/fields.py:376]: I can probably add a comment in the code in the meanwhile

Comment (sydney-runkle) [reply pydantic/fields.py:376]: @Viicos,

That'd be great. Yeah, I realize now that a lot of the changes here just showcase the utility of a `TypeForm` feature :).

**Review by sydney-runkle — COMMENTED** [review]
Awesome, happy to merge once we get the new version of `pydantic-core` in! Plus, could you add a few explanatory comments re the `cast` stuff so that the fix is explained for other folks?

**Comment by sydney-runkle** [comment]
@Viicos,

I pulled in the latest `pydantic-core`, which seems to be causing some issues...

**Review by sydney-runkle — APPROVED** [review]
I'm happy with this in its current state - I presume lots will be changed here (and in other places) once the typeform pep gets implemented :).
